### PR TITLE
Make the diffPath optional

### LINKED ISSUE #64: No output option
Would be nice to be able to not save the diffOutput. I just need the diff info :)

### PR DESCRIPTION
**Pull request by serpent7776**
Images can be compared without saving the diff file.

`Main.ml` doesn't seem to have tests, so unsure if I should add any.

Fixes #64

### CODE REVIEW

File: bin/Main.ml [hunk 1]
```diff
@@ -60,7 +60,8 @@ let main img1Path img2Path diffPath threshold outputDiffMask failOnLayoutChange
       ->
         { exitCode = 0; diff = Some diffOutput }
     | Pixel (diffOutput, diffCount, diffPercentage, _) ->
-        IO1.saveImage diffOutput diffPath;
+        if Option.is_some diffPath then
+          IO1.saveImage diffOutput @@ Option.get diffPath;
```
Comment (eWert-Online) [inline bin/Main.ml:64]: I would change this to be a `match` instead. 

```suggestion
        match diffPath with
          | None -> ()
          | Some path -> IO1.saveImage diffOutput path;
```

or even just an Option.iter:

```suggestion
        diffPath |> Option.iter (IO1.saveImage diffOutput);
```

I am on my phone currently. So its not tested and formatting might be wrong.

**Comment by eWert-Online** [comment]
odiff.js should probably also be changed 🙂

File: bin/Main.ml (hunk 1 above)
Comment (serpent7776) [reply bin/Main.ml]: I updated the code to use `Option.iter`.

**Comment by eWert-Online** [comment]
At least the type definition in npm_package/odiff.d.ts would need to be changed, so the diffPath parameter is optional. 

I guess @dmtrKovalenko can say more regarding changes to the odiff.js implementation.

**Comment by dmtrKovalenko** [comment]
Firstly thank you so much for your contribution. This is amazing!

Yes we definitely need to change the types of the odiff.d.ts file and regenerate readme. 

Also it would be good to add e2e JavaScript test to cover this functionality and check that it actually works without passing file argument.

**Comment by serpent7776** [comment]
I guess I'd need some help with the `odiff.d.ts` change. I have no idea about typescript. I don't even know what's the purpose of this file :)

**Comment by dmtrKovalenko** [comment]
Yeah I can push something tomorrow

**Comment by serpent7776** [comment]
Thanks, I updated README. Let me know if it sounds OK.

**Comment by dmtrKovalenko** [comment]
Oh yeah and after thinking a bit about this: in case there is no output we should not actually write and output buffer so we can save performance a little bit.

**Comment by serpent7776** [comment]
I made the change to not write to the `diffOutput` image if output path is not specified. 

I moved `diffOutput` image creation out of `compare` to `main`. I did it this way to avoid an extra `bool` parameter and also because `outputDiffMask` and `diffOutput` are closely related. This has caused some other changes, particularly is the tests. 
Let me know if you like this change or would rather have some other approach to this.

The difference is quite noticeable. For an 1914x7896 image it's ~862.1 ms when writing output diff and ~340.1 ms without writing.

Also, as I understand there was an issue in the case of `antialiasing = false`, because `diffOutput` was shallow copy of `img1`. So whenever `difOutput` was modified, `img1` was modified as well. Normally this doesn't matter, but for antialiased case it does, because we look back at the pixels we compared earlier. I added `clone` function to do a deep copy of the image data to fix the issue. This has caused Core antialiased tests to fail, so I refreshed expects, but now I see that the second test hasn't changed on the CI :shrug: .

There was also small issue in the TIFF tests.

**Comment by dmtrKovalenko** [comment]
> Also, as I understand there was an issue in the case of antialiasing = false, because diffOutput was shallow copy of img1. So whenever difOutput was modified, img1 was modified as well. Normally this doesn't matter, but for antialiased case it does, because we look back at the pixels we compared earlier. I added clone function to do a deep copy of the image data to fix the issue. This has caused Core antialiased tests to fail, so I refreshed expects, but now I see that the second test hasn't changed on the CI 🤷 .

this is okay because the alghoritm of checking antialiased images again calculates the changed pixel so if we already changed the pixel to the 255 255 255 the chance of it being again 255 255 255 is very small and it doesn't really harm the logic

**Comment by serpent7776** [comment]
But I did notice difference is results when specifying output and when it wasn't specified:

```
$ ./_build/install/default/bin/ODiffBin  test/test-images/aa/antialiasing-on.png test/test-images/aa/antialiasing-off-small.png --antialiasing
Failure! Images are different.
Different pixels: 417 (1.042500%)

$ ./_build/install/default/bin/ODiffBin  test/test-images/aa/antialiasing-on.png test/test-images/aa/antialiasing-off-small.png --antialiasing /tmp/diff.png
Failure! Images are different.
Different pixels: 421 (1.052500%)
```

**Comment by serpent7776** [comment]
Oh, after clean rebuild the second test hasn't changed locally I reset expect for that test.

**Comment by eWert-Online** [comment]
I am against moving the diffOutput to the main function. 

I am using odiff as a library in osnap, so i am using the diff function directly. With this change I would not be able to access the output anymore.

**Comment by serpent7776** [comment]
You are still able to access output in `?diffOutput` parameter, but yes, it's a breaking change. Waiting for different design suggestion.

**Comment by dmtrKovalenko** [comment]
yes, we should keep sticking with a good useful design for the main function as I also want to publish this is a standalone OPAM package at some point. And yes keeping it useful for @eWert-Online is a top priority. 

So my thought: what if we will make 2 top level wrapper functions like diff and diffWithoutOutput and will keep the api for diff the same but diffWithOutput will pass some kind of flag into the underlying function (yes boolean flag would be fine) and if the flag `captureDiff` is false we simply do not capturing the output

**Comment by dmtrKovalenko** [comment]
> But I did notice difference is results when specifying output and when it wasn't specified: 

Yes, I know this, but this is not a big deal in a real-world scenario, as the test is a rigorous edge case if we compare 2 images and we know exactly that one pixel was changed and it is not an antialiasing we can not be 100% that any other adjustment pixel around it is not anti-aliasing.

That's just my understanding of the anti-aliasing detection alghoritm. There might be some points I miss so would be good to see a real world image which makes the odiff to provide wrong results because of this optimization

**Comment by serpent7776** [comment]
OK, should I revert the changes related to anti-aliasing and image cloning?

**Comment by serpent7776** [comment]
@dmtrKovalenko Have you had an occasion to review the updated changes?

**Review by dmtrKovalenko — APPROVED** [review]
